README.md
# Openapi Test Suite
![npm version](https://img.shields.io/npm/v/@truesparrow/openapi-test-suite.svg?style=flat)

## Objective
This package aims to solve the following two problems:
1. Maintenance is a big problem to solve in any test suite. As the APIs evolve at a fast pace, maintenance becomes more difficult.
2. Usually when negative test cases are written, all combinations of parameters are not covered. Writing all the combinations manually is difficult and gets neglected.

We solve point 1, by consuming the openapi.json file to generate test cases in DRY manner.
Openapi.json is auto-generated using a generator script, as described in this [blog](https://truesparrow.com/blog/dry-api-docs-and-validations/).

To solve point 2, we use the cartesian product of possible (correct as well as incorrect) sets of values for each parameter to get to the set of all negative test cases.

## Approach
- For each route, we figure out the possible set of values (correct as well as incorrect) for each parameter. Now, we take the 
   cartesian product of these sets, to get all possible combinations of parameters. Correctness of the values is based 
   on the parameter type and parameter constraints given in openapi.json
- For all the incorrect combinations, we fire API calls and check whether the API response has the correct param level error.
- In each route, every optional parameter is tested with its correct, incorrect as well as null value.
- For cookie protected routes, we can include a Security Scheme object as a third argument while initializing the package. This object will be automatically consumed
   based on the specific security requirements of a route. If the cookies provided in Security Scheme object cannot be validated then http error 401 Unauthorized is expected.

## Install
For installing the npm package, run following command in terminal:
```
    npm install @truesparrow/openapi-test-suite
```

## Initialize

```node
    const openapiObj = require('@truesparrow/openapi-test-suite');
    
    const ApiTestSuite = require('@truesparrow/openapi-test-suite');
    const serverIndex = 0; // Index of the server (to be hit) from the servers block of openapi.json
    const securityScheme = require('./examples/securityScheme.json');
    const apiSuiteObj = new ApiTestSuite(openApiObj, serverIndex, securityScheme);
```

## Run Test Suite
```node
    // To run the test suite, use following command
    apiSuiteObj.runTest();

    // After the test suite run is over, it is recommended to call cleanup because the openApiObj and serverIndex are in-memory cached.
    apiSuiteObj.cleanup();
```

### Examples

#### Case 1: Expected success true but received false.
In the following, all params were passed correctly, but the API response received `success` as false.
```
    GET /api/admin/web/login/phone/otp 
    params: {"country_code":1,"raw_phone_number":"9878654325"} 
    Expected response success: true 
    Incorrect Parameters: {} 
    Response HTTP code: 200 
    API response: {
                    "success": false,
                    "err": {
                      "code": "BAD_REQUEST",
                      "msg": "Something went wrong.",
                      "error_data": [
                        {
                          "parameter": "phone_number",
                          "msg": "Invalid phone number."
                        }
                      ],
                      "internal_id": "INVALID_PHONE_NUMBER"
                    }
                  } 
    Response validation error: {"kind":"errorRespForCorrectCase"}
```

#### Case 2: Expected success false but received true.
In the following, `raw_phone_number` was passed incorrectly, but the API response received `success` as true.
```
    GET /api/admin/web/login/phone/otp 
    params: {"country_code":91,"raw_phone_number":null} 
    Expected response success: false 
    Incorrect Parameters: {"raw_phone_number":"Value cannot be null"}
    Response HTTP code: 200 
    API response: {
                    "success": true,
                    "data": {
                      "otp_detail": {
                        "id": 69,
                        "sms_identifier": "69:dfdc1b36-3fa8-4b9c-8c09-a61f334dfd9f",
                        "uts": 1653291961
                      }
                    }
                  } 
```

#### Case 3: Incorrect parameter passed but not obtained back as an error
In the following, `raw_phone_number` was passed incorrectly, but the API response `error_data` does not convey the same.
```
    GET /api/admin/web/login/phone/otp 
    params: {"country_code":"959bb2","raw_phone_number":null} 
    Expected response success: false 
    incorrectParamsMap: {"country_code":"Value cannot be string","raw_phone_number":"Value cannot be null"} 
    Response HTTP code: 200 
    API response: {
                    "success": false,
                    "err": {
                      "code": "BAD_REQUEST",
                      "msg": "Something went wrong.",
                      "error_data": [
                        {
                          "parameter": "country_code",
                          "msg": "Invalid parameter country_code.  Please ensure the input is well formed."
                        },
                        {
                          "parameter": "phone_number",
                          "msg": "Invalid phone number."
                        }
                      ],
                      "internal_id": "v_ap_rd_1"
                    }
                  } 
    Response validation error: {"kind":"parameterErrorNotObtained","parameter":"phone_number"} 
```

#### Case 4: Correct parameter passed but got API error.
In the following, all params were passed correctly, but got the api error "Something went wrong".
```
    GET /api/admin/web/login/phone/otp 
    params: {"country_code":91,"raw_phone_number":"9876543239"} 
    Expected response success: true 
    Incorrect Parameters: {} 
    Response HTTP code: 500 
    API response: {
                    "success": false,
                    "err": {
                      "code": "INTERNAL_SERVER_ERROR",
                      "msg": "Something went wrong.",
                      "error_data": [],
                      "internal_id": "PHONE_NUMBER_DOES_NOT_EXIST"
                    }
                  } 
    Response validation error: {"kind":"errorRespForCorrectCase"} 
```

#### Case 5: Correct parameter passed but error in schema validation.
In the following, all params were passed correctly, but got `respEntityTypeMismatch` error.
```
   GET /api/admin/web/login/phone/otp 
   params: {"country_code":91,"raw_phone_number":"9876543233"} 
   Expected response success: true 
   Incorrect Parameters: {} 
   Response HTTP code: 200 
   API response: {
                   "success": true,
                   "data": {
                     "otp_detail": {
                       "id": "abc",
                       "sms_identifier": "72:83c91ffc-9741-492f-8fbf-f61d6e9b9ba8",
                       "uts": 1653293902
                     }
                   }
                 } 
   Response validation error: {"kind":"respEntityTypeMismatch","debugLevel":"response.data.otp_detail.id","schemaType":"string"} 
```

#### Case 6: Incorrect cookie value and correct parameters passed.
In the following, incorrect cookie value was passed, so the expected response http code was 401; but the API response received `success` as true.
```
   GET /api/consumer/v1/user/current 
   params: {"country_code":91,"raw_phone_number":"9876543233"} 
   Expected response success: false 
   Incorrect Parameters: {} 
   Response HTTP code: 200 
   API response: {
                   "success": true,
                   "data": {
                           "current_user": {
                               "id": 100000,
                               "first_name": "ishaphone",
                               "last_name": "lastn",
                               "basic_user_detail_id": 100000,
                               "status": "ACTIVE",
                               "uts": 1651651178
                           }
                    }
                 }
   Response validation error: {"kind":"mandatoryCookieValidationFailed"}
   Request headers: {"Cookie":"aulc=25de238247286fa7fc7b9cbd071a45eda87f8690dd524cbf71c104d1d7e8252ef2dc4a698a389368dceb673238ccc47a342b611fdd00e3522183c25d6fb866faf368a7f9ec716e9df4b446d3541cd"} 
```

#### Case 7: Correct cookie value and incorrect parameters passed.
In the following, correct cookie value was passed with incorrect combination of parameters, but got http code 401.
```
    GET /api/consumer/v1/user/current 
       params: {"country_code":91,"raw_phone_number":null}
       Expected response success: false 
       Incorrect Parameters: {"raw_phone_number":"Value cannot be null"} 
       Response HTTP code: 401 
       API response: {
                       "success": false,
                       "err": {
                         "code": "UNAUTHORIZED",
                         "msg": "Access denied due to invalid credentials.",
                         "error_data": [],
                         "internal_id": "l_ch_vclcr_2"
                       }
                     }
       Response validation error: {"kind":"unauthorizedApiRequestForValidCookie"}
       Request headers: {"Cookie":"aulc=25de238247286fa7fc7b9cbd071a45eda87f8690dd524cbf71c104d1d7e8252ef2dc4a698a389368dceb673238ccc47a342b611fdd00e3522183c25d6fb866faf368a7f9ec716e9df4b446d3541cd"} 
```

## Schema Validator
Using schema validator, the data can be recursively cross validated against its schema.

### Examples
Get schema validator class using apiSuiteObj.
```
    const schemaValidator = apiSuiteObj.SchemaValidator;
```
Then call validateObjectBySchema(data,dataSchema,debugLevel) method of schemaValidator class with following arguments:
1) data = Data to be validated (dataType : object).
2) dataSchema = Schema against which the given data should be validated. (dataType : object).
3) debugLevel = String value which helps in debugging (dataType : string).

Note: validateObjectBySchema function throws error only in case of incorrect data.
#### Case 1: Validate object data.
```
    const data = require('./examples/dataToBeValidated_1.json');
    const dataSchema = require('./examples/dataSchema_1.json');
    const correctData = data.correctData;
    const incorrectData = data.incorrectData;
```
With correct data
```
new schemaValidator().validateObjectBySchema(correctData,dataSchema,'Response');
```
With incorrect data
```
new schemaValidator().validateObjectBySchema(incorrectData,dataSchema,'Response');
```
Logs:
```
{
  kind: 'respEntityTypeMismatch',
  debugLevel: 'VALUE(Response).status',
  schemaType: 'string'
}
```
Here kind indicates the type of validation error, debugLevel indicates the level at which the validation failed and schemaType indicates expected schema.
#### Case 2: Validate array of object data.
```
    const data = require('./examples/dataToBeValidated_2.json');
    const dataSchema = require('./examples/dataSchema_2.json');
    const correctData = data.correctData;
    const incorrectData = data.incorrectData;
```
```
new schemaValidator().validateObjectBySchema(incorrectData,dataSchema,'Response');
```
Logs:
```
{
  kind: 'respEntityTypeMismatch',
  debugLevel: 'Response.[0].name',
  schemaType: 'string'
}
```
#### Case 3: Validate array of integer data.
```
    const data = require('./examples/dataToBeValidated_3.json');
    const dataSchema = require('./examples/dataSchema_3.json');
    
    const correctData = data.correctData;
    const incorrectData = data.incorrectData;
```
```
new schemaValidator().validateObjectBySchema(incorrectData,dataSchema,'Response');
```
Logs:
```
{
  kind: 'respEntityTypeMismatch',
  debugLevel: 'Response',
  schemaType: 'array'
}
```
#### Case 4: Validate string data.
```
    const data = require('./examples/dataToBeValidated_4.json');
    const dataSchema = require('./examples/dataSchema_4.json');
    
    const correctData = data.correctData;
    const incorrectData = data.incorrectData;
```
```
new schemaValidator().validateObjectBySchema(incorrectData,dataSchema,'Response');
```
Logs:
```
{
  kind: 'respEntityTypeMismatch',
  debugLevel: 'Response',
  schemaType: 'string'
}
```
## SQL & Executable Commands Injection.
Using Start security test you can test your API parameters' vulnerabilities against SQL and Command Injection attacks.

Manual monitoring of API logs is needed to check if any resources are affected by SQL or Commands injection queries.

Execute security test class using apiSuiteObj.
```
    apiSuiteObj.runSecurityTest();
```
### Example
API request will be executed for injection queries against each parameter.

```
GET /api/consumer/v1/signup/phone/otp 
Request headers: {} 
params: {"country_code":"OR 1=1#","raw_phone_number":"1111111111"}
```


## Future Scope
- Error schema in component section and it's use.


CHANGELOG.md
## Openapi-test-suite v2.0.0
- Republished package under @truesparrow

## Openapi-test-suite v1.0.1
- LICENSE changes.

## Openapi-test-suite v1.0.0
This is the very first release of this package.
- runTests method added which starts the DRY test suite based on openapi.json file.
- SchemaValidator exposed to validate any data vis-a-vis it's schema.
- Running tests for SQL injections and executable command injections.


lib/Suite/Start.js
const rootPrefix = '../..',
  HttpLibrary = require(rootPrefix + '/lib/HttpRequest'),
  helper = require(rootPrefix + '/lib/helper'),
  configProvider = require(rootPrefix + '/lib/configProvider'),
  ParamValueSetFactory = require(rootPrefix + '/lib/paramValueSet/Factory'),
  logger = require(rootPrefix + '/lib/logger/customConsoleLogger'),
  CommonValidator = require(rootPrefix + '/lib/validators/Common'),
  util = require(rootPrefix + '/lib/util');

/**
 * Class for starting openapi-test-suite
 *
 * @class Start
 */
class Start {
  /**
   * Constructor
   *
   * @constructor
   */
  constructor(params) {
    const oThis = this;

    oThis.openapiObj = configProvider.getConfig('openapiObj');
    oThis.serverIndex = configProvider.getConfig('serverIndex');

    oThis.serverUrl = oThis.openapiObj.definition.servers[oThis.serverIndex].url;
    oThis.allPaths = oThis.openapiObj.definition.paths;
    oThis.globalSecurity = oThis.openapiObj.definition.security || {};
    oThis.securitySchemes = oThis.openapiObj.definition.components.securitySchemes;
    oThis.securityInfo = configProvider.getConfig('securityInfo');
  }

  /**
   * Main performer for class.
   *
   * @returns {Promise<void>}
   */
  async perform() {
    const oThis = this;

    await oThis.testAllApis();
  }

  /**
   * Test all apis
   *
   * @returns {Promise<void>}
   */
  async testAllApis() {
    const oThis = this;

    logger.log('--------------------------- STARTED TESTING APIS --------------------------- ');

    for (const path in oThis.allPaths) {
      const currentPath = oThis.allPaths[path];
      for (const method in currentPath) {
        const methodData = currentPath[method],
          methodParams = methodData.parameters,
          methodSecurity = methodData.security || oThis.globalSecurity;

        const paramsCombinations = await oThis.generateApiParamsCombination(methodParams),
          incorrectCombinations = paramsCombinations.incorrectCombinations,
          correctCombinations = paramsCombinations.correctCombinations;

        if (CommonValidator.validateNonEmptyObject(methodSecurity)) {
          // To test each security auth test cases with valid api params
          const sampleCorrectCombination = correctCombinations[0];

          for (const securityAuth in methodSecurity) {
            const securityScheme = oThis.securitySchemes[securityAuth];
            if (!securityScheme) {
              throw new Error('No security scheme found for - ' + securityAuth);
            }
            const securityAuthObj = oThis.securityInfo[securityAuth] || null;

            if (!securityAuthObj) {
              throw new Error('No security auth found for - ' + securityAuth);
            }

            // Test with cookie, headers and correct api param combination. Expected http code other than 401.
            await oThis.sendRequestAndValidateResponse(
              method,
              path,
              sampleCorrectCombination,
              {
                areAllParamsCorrect: true,
                areCookiesIncorrect: false
              },
              securityAuthObj.cookies,
              securityAuthObj.headers
            );

            // Test with cookie, headers and incorrect api param combinations. Expected http code other than 401 and proper param level error.
            for (let combinationIndex = 0; combinationIndex < incorrectCombinations.length; combinationIndex++) {
              const paramsElement = incorrectCombinations[combinationIndex];
              await oThis.sendRequestAndValidateResponse(
                method,
                path,
                paramsElement,
                {
                  areAllParamsCorrect: false,
                  areCookiesIncorrect: false
                },
                securityAuthObj.cookies,
                securityAuthObj.headers
              );
            }

            const invalidCookiesMap = oThis.generateInvalidCookieValues(securityAuthObj.cookies);

            // Test with invalid cookie values, headers and correct api param combination. Expected http code 401.
            await oThis.sendRequestAndValidateResponse(
              method,
              path,
              sampleCorrectCombination,
              {
                areAllParamsCorrect: true,
                areCookiesIncorrect: true
              },
              invalidCookiesMap,
              securityAuthObj.headers
            );
          }

          // Test with headers, correct api param combination and no cookies. Expected http code 401.
          await oThis.sendRequestAndValidateResponse(method, path, sampleCorrectCombination, {
            areAllParamsCorrect: true,
            areCookiesIncorrect: true
          });
        } else {
          // Test all incorrect api param combinations, expect proper param level error.
          for (let combinationIndex = 0; combinationIndex < incorrectCombinations.length; combinationIndex++) {
            const paramsElement = incorrectCombinations[combinationIndex];
            await oThis.sendRequestAndValidateResponse(method, path, paramsElement, {
              areAllParamsCorrect: false,
              areCookiesIncorrect: false
            });
          }
        }
      }
      logger.log('--------------------------- ENDED TESTING APIS ---------------------------');
    }
  }

  /**
   * Send request and validate response.
   *
   * @param {string} method
   * @param {string} path
   * @param {object} paramsCombination
   * @param {object} validationMeta
   * @param {object} [cookies]
   * @param {object} [headers]
   * @returns {Promise<void>}
   */
  async sendRequestAndValidateResponse(
    method,
    path,
    paramsCombination,
    validationMeta,
    cookies = null,
    headers = null
  ) {
    const oThis = this;
    const reqHeaders = Object.assign({}, headers);

    if (CommonValidator.validateNonEmptyObject(cookies)) {
      reqHeaders['Cookie'] = oThis.getCookieString(cookies);
    }

    const apiParams = paramsCombination.apiParams;
    const params = {
      serverUrl: oThis.serverUrl,
      path: path,
      method: method,
      apiParameters: apiParams,
      header: CommonValidator.validateNonEmptyObject(reqHeaders) ? reqHeaders : null
    };

    const responseData = await new HttpLibrary(params).perform();
    await helper.sleep(500);

    try {
      const dataSchema = oThis.allPaths[path][method].responses['200'].content['application/json'].schema;

      await oThis.validateResponse(
        responseData,
        validationMeta.areAllParamsCorrect,
        paramsCombination.incorrectParamsMap || {},
        dataSchema,
        validationMeta.areCookiesIncorrect
      );
    } catch (error) {
      logger.error('---------------------------');
      logger.error(method.toUpperCase(), path);
      logger.error('params:', apiParams);
      logger.error(
        'Expected response success:',
        validationMeta.areAllParamsCorrect && !validationMeta.areCookiesIncorrect
      );
      logger.error('Incorrect Parameters:', paramsCombination.incorrectParamsMap || {});
      logger.error('Response HTTP code:', responseData.data.httpCode);
      logger.error('API response:', responseData.data.apiResponse);
      logger.error('Response validation error:', error);
      logger.error('Request headers:', reqHeaders);
    }
  }

  /**
   * Validate response.
   *
   * @param {object} responseData
   * @param {boolean} areAllParamsCorrect
   * @param {object} incorrectParamsMap
   * @param {object} dataSchema
   * @param {boolean} areCookiesIncorrect
   *
   * @returns {Promise<void>}
   */
  async validateResponse(responseData, areAllParamsCorrect, incorrectParamsMap, dataSchema, areCookiesIncorrect) {
    const oThis = this,
      apiResponse = responseData.data.apiResponse,
      statusCode = responseData.data.httpCode;

    if (areAllParamsCorrect) {
      if (!areCookiesIncorrect && statusCode === 401) {
        throw {
          kind: 'unauthorizedApiRequestForValidCookie'
        };
      }

      if (areCookiesIncorrect && statusCode != 401) {
        throw {
          kind: 'mandatoryCookieValidationFailed'
        };
      }
    } else {
      if (!areCookiesIncorrect && statusCode == 401) {
        throw {
          kind: 'unauthorizedApiRequestForValidCookie'
        };
      }
      const errorData = apiResponse.err.error_data;
      for (const error of errorData) {
        const parameter = error.parameter;
        if (CommonValidator.isVarNullOrUndefined(incorrectParamsMap[parameter])) {
          throw {
            kind: 'parameterErrorNotObtained',
            parameter: parameter
          };
        }
      }
    }
  }

  /**
   * Method to generate combinations of api params.
   *
   * @param {array} params
   * @returns {Promise<void>}
   */
  async generateApiParamsCombination(params) {
    const oThis = this;

    const paramNames = [],
      allParamsPossibleValuesArray = [],
      paramNameToValidValues = {},
      incorrectApiParamsCombinations = [],
      correctApiParamsCombinations = [];

    if (!params || params.length == 0) {
      return { incorrectCombinations: incorrectApiParamsCombinations, correctCombinations: [{ apiParams: {} }] };
    }

    for (const paramObj of params) {
      const schema = paramObj.schema,
        paramName = paramObj.name;
      paramNames.push(paramName);

      const schemaTypeObj = new ParamValueSetFactory().getInstance(schema, paramObj.required),
        paramCorrectValues = schemaTypeObj.correctValues(),
        paramIncorrectValues = schemaTypeObj.incorrectValues();

      paramNameToValidValues[paramName] = paramCorrectValues;
      const possibleParamValues = paramCorrectValues.concat(paramIncorrectValues);

      allParamsPossibleValuesArray.push(possibleParamValues);
    }

    const allCombinations = helper.cartesianProduct(allParamsPossibleValuesArray);

    for (let index = 0; index < allCombinations.length; index++) {
      const paramsCombination = allCombinations[index],
        apiParams = {};

      let areAllParamsCorrect = true;
      const incorrectParamsMap = {};
      for (let paramIndex = 0; paramIndex < paramNames.length; paramIndex++) {
        const paramName = paramNames[paramIndex];
        const paramValue = paramsCombination[paramIndex].value,
          paramDescription = paramsCombination[paramIndex].description;
        apiParams[paramName] = paramValue;
        const correctValuesArray = paramNameToValidValues[paramName];

        const checkValue = (obj) => obj.value === paramValue;

        if (!correctValuesArray.some(checkValue)) {
          incorrectParamsMap[paramName] = paramDescription;
          areAllParamsCorrect = false;
        }
      }

      if (areAllParamsCorrect) {
        correctApiParamsCombinations.push({
          apiParams: apiParams
        });
      } else {
        incorrectApiParamsCombinations.push({
          apiParams: apiParams,
          incorrectParamsMap: incorrectParamsMap
        });
      }
    }

    return { incorrectCombinations: incorrectApiParamsCombinations, correctCombinations: correctApiParamsCombinations };
  }

  /**
   * Get cookie string for request headers.
   *
   * @param {object} cookiesMap
   *
   * @returns {string}
   */
  getCookieString(cookiesMap) {
    const oThis = this;

    const cookieResponseArr = [];

    for (const cookieName in cookiesMap) {
      const cookieValue = cookiesMap[cookieName];
      const cookieString = cookieName + '=' + cookieValue;
      cookieResponseArr.push(cookieString);
    }

    return cookieResponseArr.join('; ');
  }

  /**
   * Get invalid cookie values.
   *
   * @param {object} cookiesMap
   *
   * @returns {object}
   */
  generateInvalidCookieValues(cookiesMap) {
    const oThis = this;

    const cookieResponseMap = {};
    for (const cookieName in cookiesMap) {
      cookieResponseMap[cookieName] = util.getRandomString(cookiesMap[cookieName].length);
    }

    return cookieResponseMap;
  }
}

module.exports = Start;
